fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1409,12 +1424,21 @@ def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
                     else:
                         m.assert_state(TrainingState.BACKWARD_PRE)
                 else:
-                    # Unlikely case. When `m` and its children has no params or has params but
-                    # none with `requires_grad==True`, then m's pre-backward and post-backward
-                    # hooks aren't called by autograd. Therefore, it is in IDLE state.
-                    m.assert_state(TrainingState.IDLE)
+                    # When `m` and its children has no params or has params but
+                    # none with `requires_grad==True`, there are two cases:
+                    # 1. output tensors are `requires_grad==True`. In this case,
+                    # pre-backward hook is still registered, so it is in BACKWARD_PRE state.
+                    # 2. output tensors are `requires_grad==False`. In this case,
+                    # pre-backward hook is not registered, so it is in IDLE state.
+                    m.assert_state([TrainingState.BACKWARD_PRE, TrainingState.IDLE])
                 m.training_state = TrainingState.IDLE
 
+                if m._is_root:
+                    # reset this flag for cases like "one forward pass + multiple backward passes"
+                    self._post_backward_callback_queued = False
+                    # indicate this post backward callback was fired
+                    self._post_backward_callback_fired = True
```

Comment:
again, I think the flag might be replaceable by asserting on the TrainingState? Perhaps need to add one more state for before and after the final call back?

Reply:
right now, trainingState was reset to be IDLE after final callback is done, let me see whether we can add a new state after the final call back

Reply:
had a second thought, I think it makes more sense that trainingState was reset to be IDLE when final callback is done, as many other functions in the codes rely on checking 'IDLE' state to make sure those functions are not called in a forward/backward pass. so it is better to reset state to be IDLE when a pass is done. 

That being said, it is probably better to use flag here.

Reply:
Cool. Let me know if this can be done. I think one less bool for checking will help make the code more readable and maintainable.

Reply:
had a second thought, I think it makes more sense that trainingState was reset to be IDLE when final callback is done, as many other functions in the codes rely on checking 'IDLE' state to make sure those functions are not called in a forward/backward pass. so it is better to reset state to be IDLE when a pass is done.

That being said, it is probably better to use flag here.